lib/libc/sys/sigprocmask.2
.\"	BSDI sigprocmask.2,v 2.1 1995/02/03 06:51:37 polk Exp
.\"
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)sigprocmask.2	8.1 (Berkeley) 6/4/93
.\"
.Dd June 4, 1993
.Dt SIGPROCMASK 2
.Os
.Sh NAME
.Nm sigprocmask
.Nd manipulate current signal mask
.Sh SYNOPSIS
.Fd #include <signal.h>
.Ft int
.Fn sigprocmask "int how" "const sigset_t *set" "sigset_t *oset"
.Sh DESCRIPTION
The
.Fn sigprocmask
function examines and/or changes the current signal mask (those signals
that are blocked from delivery).
Signals are blocked if they are members of the current signal mask set.
.Pp
If
.Fa set
is not null, the action of
.Fn sigprocmask
depends on the value of the parameter
.Fa how .
The signal mask is changed as a function of the specified
.Fa set
and the current mask.
The function is specified by
.Fa how
using one of the following values from
.Aq Pa signal.h :
.Bl -tag -width SIG_UNBLOCK
.It Dv SIG_BLOCK
The new mask is the union of the current mask and the specified
.Fa set .
.It Dv SIG_UNBLOCK
The new mask is the intersection of the current mask
and the complement of the specified
.Fa set .
.It Dv SIG_SETMASK
The current mask is replaced by the specified
.Fa set .
.El
.Pp
If
.Fa oset
is not null, it is set to
the previous value of the signal mask.
When
.Fa set
is null,
the value of
.Ar how
is insignificant and the mask remains unset
providing a way to examine the signal mask without modification.
.Pp
The system
quietly disallows
.Dv SIGKILL
or
.Dv SIGSTOP
to be blocked.
.Sh RETURN VALUES
A 0 value indicated that the call succeeded.  A -1 return value
indicates an error occurred and
.Va errno
is set to indicated the reason.
.Sh ERRORS
The
.Fn sigprocmask
call will fail and the signal mask will be unchanged if one
of the following occurs:
.Bl -tag -width Bq Er EINVAL
.It Bq Er EINVAL
.Fa how
has a value other than those listed here.
.Sh SEE ALSO
.Xr kill 2 ,
.Xr sigaction 2 ,
.Xr sigsetops 3 ,
.Xr sigsuspend 2
.Sh STANDARDS
The
.Nm sigprocmask
function call is expected to
conform to
.St -p1003.1-88 .


lib/libc/sparc/sys/sigprocmask.s
/*
 * Copyright (c) 1992, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This software was developed by the Computer Systems Engineering group
 * at Lawrence Berkeley Laboratory under DARPA contract BG 91-66 and
 * contributed to Berkeley.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * from: /master/lib/libc/sparc/sys/sigprocmask.s,v 2.1 1995/02/03 06:40:03 polk Exp
 */

#if defined(LIBC_SCCS) && !defined(lint)
	.asciz "@(#)sigprocmask.s	8.1 (Berkeley) 6/4/93"
#endif /* LIBC_SCCS and not lint */

#include "SYS.h"

/*
 * sigprocmask(int how, sigset_t *set, sigset_t *oset)
 */
ENTRY(sigprocmask)
	tst	%o1		! set == NULL?
	bne,a	1f		! if not,
	 ld	[%o1], %o1	!    replace it in %o1 with *set
!	clr	%o1		! else block no signals ...
	mov	1, %o0		! ... using sigprocmask(SIG_BLOCK)
1:
	mov	SYS_sigprocmask, %g1
	t	ST_SYSCALL
	bcc	2f		! if success,
	 tst	%o2		!    check to see if oset requested
	ERROR()
2:
	bne,a	3f		! if oset != NULL,
	 st	%o0, [%o2]	!    *oset = oldmask
3:
	retl			! in any case, return 0
	 clr	%o0
